sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/Combine.java

```diff
@@ -124,14 +124,16 @@ private Combine() {
     return globally(fn, displayDataForFn(fn));
   }
 
-  private static <T> DisplayData.ItemSpec<? extends Class<?>> displayDataForFn(T fn) {
-    return DisplayData.item("combineFn", fn.getClass())
+  private static DisplayData.ItemSpecBase<?, ?> displayDataForFn(Object fn) {
```

Comment:
3 is ok but it requires a person to know about the simple class wrapping display data. It is an easy way to reduce the amount of boiler plate. It also allows other wrappers to be produced.

1 is ok as well since inferTypes is meant to guess while everything else is explicit

I don't like 2 since the "class" may not always be the item that you want to use and only the caller would know requiring a potentially awkward follow up call that requires you to override what was done automatically.